libyang/extension.py

```diff
@@ -0,0 +1,178 @@
+
+
+@ffi.def_extern(name="lypy_lyplg_ext_parse_clb")
+def libyang_c_lyplg_ext_parse_clb(pctx, pext):
+    plugin = extensions_plugins[pext.record.plugin]
+    module_cdata = lib.lyplg_ext_parse_get_cur_pmod(pctx).mod
+    context = Context(cdata=module_cdata.ctx)
+    module = Module(context, module_cdata)
+    parsed_ext = ExtensionParsed(context, pext, module)
+    plugin.set_parse_ctx(pctx)
+    try:
+        plugin.parse_clb(module, parsed_ext)
+        return lib.LY_SUCCESS
+    except LibyangExtensionError as e:
+        ly_level = get_libyang_level(e.log_level)
+        if ly_level is None:
+            ly_level = lib.LY_EPLUGIN
+        e_str = str(e)
+        plugin.add_error_message(e_str)
+        lib.lyplg_ext_parse_log(pctx, pext, ly_level, e.ret, str2c(e_str))
+        return e.ret
+
+
+@ffi.def_extern(name="lypy_lyplg_ext_compile_clb")
+def libyang_c_lyplg_ext_compile_clb(cctx, pext, cext):
+    plugin = extensions_plugins[pext.record.plugin]
+    context = Context(cdata=lib.lyplg_ext_compile_get_ctx(cctx))
+    module = Module(context, cext.module)
+    parsed_ext = ExtensionParsed(context, pext, module)
+    compiled_ext = ExtensionCompiled(context, cext)
+    plugin.set_compile_ctx(cctx)
+    try:
+        plugin.compile_clb(parsed_ext, compiled_ext)
+        return lib.LY_SUCCESS
+    except LibyangExtensionError as e:
+        ly_level = get_libyang_level(e.log_level)
+        if ly_level is None:
+            ly_level = lib.LY_EPLUGIN
+        e_str = str(e)
+        plugin.add_error_message(e_str)
+        lib.lyplg_ext_compile_log(cctx, cext, ly_level, e.ret, str2c(e_str))
+        return e.ret
+
+
+@ffi.def_extern(name="lypy_lyplg_ext_parse_free_clb")
+def libyang_c_lyplg_ext_parse_free_clb(ctx, pext):
+    plugin = extensions_plugins[pext.record.plugin]
+    context = Context(cdata=ctx)
+    parsed_ext = ExtensionParsed(context, pext, None)
+    plugin.parse_free_clb(parsed_ext)
+
+
+@ffi.def_extern(name="lypy_lyplg_ext_compile_free_clb")
+def libyang_c_lyplg_ext_compile_free_clb(ctx, cext):
+    plugin = extensions_plugins[getattr(cext, "def").plugin]
+    context = Context(cdata=ctx)
+    compiled_ext = ExtensionCompiled(context, cext)
+    plugin.compile_free_clb(compiled_ext)
+
+
+class ExtensionPlugin:
+    ERROR_SUCCESS = lib.LY_SUCCESS
+    ERROR_MEM = lib.LY_EMEM
+    ERROR_INVALID_INPUT = lib.LY_EINVAL
+    ERROR_NOT_VALID = lib.LY_EVALID
+    ERROR_DENIED = lib.LY_EDENIED
+    ERROR_NOT = lib.LY_ENOT
+
+    def __init__(
+        self,
+        module_name: str,
+        name: str,
+        id_str: str,
+        context: Optional[Context] = None,
+        parse_clb: Optional[Callable[[Module, ExtensionParsed], None]] = None,
+        compile_clb: Optional[
+            Callable[[ExtensionParsed, ExtensionCompiled], None]
+        ] = None,
+        parse_free_clb: Optional[Callable[[ExtensionParsed], None]] = None,
+        compile_free_clb: Optional[Callable[[ExtensionCompiled], None]] = None,
+    ) -> None:
```

Comment:
We need a docstring to explain the arguments and callback formats.

Reply:
Done

Reply:
Done